### PR TITLE
get beacon committees endpoint

### PR DESCRIPTION
**Pull request by rolfyone**
add /beacon/committees endpoint to allow committeeAssignments to be returned.

added missing test to show no content for /beacon/state

Updated readme which was missing a link to the spec.

Addresses #1197

### CODE REVIEW

**Comment by rolfyone** [comment]
so a comment on functionality.
- can query back as far as slot 0
- can't query 'next' epoch, meaning you need to wait until the epoch starts - that might be a problem...

File: data/beaconrestapi/src/main/java/tech/pegasys/artemis/beaconrestapi/beaconhandlers/BeaconCommitteesHandler.java [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+      method = HttpMethod.GET,
+      summary = "Get the committees for all slots and committee indices in a given epoch.",
+      tags = {TAG_BEACON},
+      queryParams = {
+        @OpenApiParam(
+            name = EPOCH,
+            description = "The epoch for which committees will be returned.",
+            required = true),
+      },
+      description = "Request a list of the Committee assignments for a given epoch.",
```
Comment (macfarla) [inline data/beaconrestapi/src/main/java/tech/pegasys/artemis/beaconrestapi/beaconhandlers/BeaconCommitteesHandler.java]:
```suggestion
      description = "Request a list of the committee assignments for a given epoch.",
```

File: data/beaconrestapi/src/main/java/tech/pegasys/artemis/beaconrestapi/beaconhandlers/BeaconCommitteesHandler.java [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+  public BeaconCommitteesHandler(
+      CombinedChainDataClient combinedClient, JsonProvider jsonProvider) {
+    this.combinedClient = combinedClient;
+    this.jsonProvider = jsonProvider;
+  }
+
+  @OpenApi(
+      path = ROUTE,
+      method = HttpMethod.GET,
+      summary = "Get the committees for all slots and committee indices in a given epoch.",
```
Comment (macfarla) [inline data/beaconrestapi/src/main/java/tech/pegasys/artemis/beaconrestapi/beaconhandlers/BeaconCommitteesHandler.java]: committee assignments?

Comment (macfarla) [reply data/beaconrestapi/src/main/java/tech/pegasys/artemis/beaconrestapi/beaconhandlers/BeaconCommitteesHandler.java]: will people reading this know what committee indices are?

Comment (ajsutton) [reply data/beaconrestapi/src/main/java/tech/pegasys/artemis/beaconrestapi/beaconhandlers/BeaconCommitteesHandler.java]: Committee indices is the correct term from the spec.  A committee assignment is the set of committee indices in the committee.

Comment (rolfyone) [reply data/beaconrestapi/src/main/java/tech/pegasys/artemis/beaconrestapi/beaconhandlers/BeaconCommitteesHandler.java]: so should i leave this comment as-is? or change the 'get the committees' to 'get the committee assignments'?

File: storage/src/main/java/tech/pegasys/artemis/storage/CombinedChainDataClient.java [hunk 3]
```diff
@@ -178,4 +185,49 @@ private boolean isFinalized(final UnsignedLong slot) {
...
+  public SafeFuture<List<CommitteeAssignment>> getCommitteeAssignmentAtEpoch(UnsignedLong epoch) {
+    final UnsignedLong startingSlot = compute_start_slot_at_epoch(epoch);
+
+    SafeFuture<Optional<BeaconState>> future =
+        getStateAtSlot(startingSlot, recentChainData.getBestBlockRoot());
+    if (future != null) {
+      return future
+          .thenCompose(
+              optionalState ->
+                  completedFuture(getCommitteesFromStateWithCache(optionalState, startingSlot)))
```
Comment (ajsutton) [inline storage/src/main/java/tech/pegasys/artemis/storage/CombinedChainDataClient.java]: You can just use `thenApply` here and not need to wrap in a `completedFuture`.

File: storage/src/test/java/tech/pegasys/artemis/storage/CombinedChainDataClientTest.java [hunk 4]
```diff
@@ -231,6 +235,16 @@ public void getBlockInEffectAtSlot_returnPreviousBlockWhenSlotWasEmpty() {
...
   }
 
+  @Test
+  public void getCommitteesFromStateWithCache_shouldReturnCommitteeAssignments() {
+    BeaconStateWithCache stateWithCache =
+        BeaconStateWithCache.fromBeaconState(DataStructureUtil.randomBeaconState(11233));
+    List<CommitteeAssignment> data =
+        client.getCommitteesFromStateWithCache(
+            Optional.of(stateWithCache), stateWithCache.getSlot());
+    assertEquals(8, data.size());
```
Comment (mark-terry) [inline storage/src/test/java/tech/pegasys/artemis/storage/CombinedChainDataClientTest.java]: AssertJ?

File: storage/src/main/java/tech/pegasys/artemis/storage/CombinedChainDataClient.java [hunk 5]
```diff
@@ -178,4 +185,49 @@ private boolean isFinalized(final UnsignedLong slot) {
...
+   *
+   * @param epoch - the current or historic epoch
+   * @return list of CommitteeAssignments
+   */
+  public SafeFuture<List<CommitteeAssignment>> getCommitteeAssignmentAtEpoch(UnsignedLong epoch) {
+    final UnsignedLong startingSlot = compute_start_slot_at_epoch(epoch);
+
+    SafeFuture<Optional<BeaconState>> future =
+        getStateAtSlot(startingSlot, recentChainData.getBestBlockRoot());
+    if (future != null) {
```
Comment (mark-terry) [inline storage/src/main/java/tech/pegasys/artemis/storage/CombinedChainDataClient.java]: When is this null?  If the store is null?

**Review by mark-terry — COMMENTED** [review]
A couple of queries.

File: storage/src/main/java/tech/pegasys/artemis/storage/CombinedChainDataClient.java [hunk 6]
```diff
@@ -178,4 +185,49 @@ private boolean isFinalized(final UnsignedLong slot) {
...
+   * Gets a list of CommitteeAssignments for a specified epoch.
+   *
+   * @param epoch - the current or historic epoch
+   * @return list of CommitteeAssignments
+   */
+  public SafeFuture<List<CommitteeAssignment>> getCommitteeAssignmentAtEpoch(UnsignedLong epoch) {
+    final UnsignedLong startingSlot = compute_start_slot_at_epoch(epoch);
+
+    SafeFuture<Optional<BeaconState>> future =
+        getStateAtSlot(startingSlot, recentChainData.getBestBlockRoot());
```
Comment (ajsutton) [inline storage/src/main/java/tech/pegasys/artemis/storage/CombinedChainDataClient.java:208]: Committees are always calculated for one epoch ahead.  So if you want to calculate committees for `epoch` you need to get the state for `epoch - 1`.  The exception is epoch 0 which uses genesisState (as does epoch 1).

So you are actually able to calculate committees up to one epoch ahead of the current one.

**Review by ajsutton — COMMENTED** [review]
Looks good, just got to ensure we use the right state when calculating.

File: data/beaconrestapi/src/main/java/tech/pegasys/artemis/beaconrestapi/beaconhandlers/BeaconCommitteesHandler.java (hunk 2 above)
Comment (ajsutton) [reply data/beaconrestapi/src/main/java/tech/pegasys/artemis/beaconrestapi/beaconhandlers/BeaconCommitteesHandler.java]: It's probably fine as-is but suspect `Get the committee assignments for a given epoch` is all that's actually required and probably the most accurate.  @benjaminion ?

File: storage/src/main/java/tech/pegasys/artemis/storage/CombinedChainDataClient.java (hunk 3 above)
Comment (rolfyone) [reply storage/src/main/java/tech/pegasys/artemis/storage/CombinedChainDataClient.java]: sounds less messy

File: storage/src/main/java/tech/pegasys/artemis/storage/CombinedChainDataClient.java (hunk 6 above)
Comment (rolfyone) [reply storage/src/main/java/tech/pegasys/artemis/storage/CombinedChainDataClient.java:208]: ok i've done that and now i think our testing may be inadequate as no tests broke, suggesting that there's too much mocking to see this pretty significant change.

File: storage/src/main/java/tech/pegasys/artemis/storage/CombinedChainDataClient.java [hunk 7]
```diff
@@ -178,6 +187,47 @@ private boolean isFinalized(final UnsignedLong slot) {
...
+  }
+
+  /**
+   * Gets a list of CommitteeAssignments for a specified epoch.
+   *
+   * @param epoch - the current or historic epoch
+   * @return list of CommitteeAssignments
+   */
+  public SafeFuture<List<CommitteeAssignment>> getCommitteeAssignmentAtEpoch(UnsignedLong epoch) {
+    final UnsignedLong committeesCalculatedAtEpoch = epoch.equals(ZERO) ? ZERO : epoch.minus(ONE);
```
Comment (macfarla) [inline storage/src/main/java/tech/pegasys/artemis/storage/CombinedChainDataClient.java:204]:
```suggestion
    final UnsignedLong epochToCalculateCommittees = epoch.equals(ZERO) ? ZERO : epoch.minus(ONE);
```

Comment (macfarla) [reply storage/src/main/java/tech/pegasys/artemis/storage/CombinedChainDataClient.java:204]: names are hard. still don't love it

**Review by macfarla — COMMENTED** [review]
some minor comments

File: storage/src/main/java/tech/pegasys/artemis/storage/CombinedChainDataClient.java (hunk 7 above)
Comment (macfarla) [reply storage/src/main/java/tech/pegasys/artemis/storage/CombinedChainDataClient.java:204]: also I don't get why we need to get committees at epoch - 1. Is it self-explanatory or do we need to make it clear?

**Review by macfarla — APPROVED** [review]

**Pull request merged**